cilium/CFP-27752-operator-manages-cilium-identities.md

```diff
@@ -0,0 +1,214 @@
+
+The diagram below covers high level actions that happen when a pod is created. From pod scheduled to pod ready, and to CEP and CID data being received by all nodes in the cluster. CEP batching into CES is included, but that’s an optional step based on configuration. Comparison before (red arrows) and after (blue arrows) changes.
+
+![](../cilium/images/CFP-27752-pod-creation-process.png)
+
+
+CEP creation in parallel triggers its endpoint regeneration on the host node and CEP propagation to all other nodes.
+
+Endpoint regeneration happens when:
+
+* An endpoint is created and updated, including pod and namespace label changes (security identity change).
+* CID is created or deleted - all Agents regenerate all of their endpoints.
+* Network policies are changed.
+
+IPCache is updated when remote CEPs/CESs are received.
+
+## Label changes
+
+
+### Pod
+
+Pod/Deployment label changes result in pod recreation, which follows the same pod creation path previously described.
+
+
+### Namespace
+
+When namespace labels change, Operator and Agents get the update at the same time. Agents wait until Operator processes the namespace label change update and creates all the new CIDs. Agents keep retrying to match CIDs for all their local endpoints until they are received through the CID watcher.
+
+
+## Simultaneous CID management compatibility
+
+The only problem with simultaneous CID management (both Operator and Agents are managing CIDs) is when namespace labels are changed. The only bad thing that happens is that there is CID duplication. This is an existing problem already, that doesn't cause disruption.
+
+Simultaneous CID management is only expected to happen during upgrade from Agents managing CIDs to Operator managing CIDs.
+
+
+## CID deduplication
+
+Operator will not create duplicate CIDs, but it will allow for their existence. In case other sources, like Agents, are making duplicate CIDs, Operator will be compatible with tracking multiple CIDs. This is needed to make Operator compatible with Agents managing CIDs at the same time.
+
+
+## KVStore compatibility
+
+[KVStore](https://docs.cilium.io/en/stable/kvstore/) is a key-value store that is mostly used for sharing Cilium mappings between IPs, identities (IPCache) and labels, across the cluster ([what's inside Cilium’s KVStore](http://arthurchiao.art/blog/whats-inside-cilium-etcd/)).
+
+Operator managing CIDs is intended to work with both `identity-allocation-mode=kvstore` and `identity-allocation-mode=crd`. CRD being the default, the initial implementation will use the CRD allocation mode, and later expanded to other allocation modes.
+
+All features that rely on using kvstore will not work until CID controller is expanded to work with kvstore allocation mode. The work to enable kvstore compatibility is estimated to of medium size and it will be tracked in a separate GitHub issue (https://github.com/cilium/cilium/issues/34865).
```

Comment:
What exactly does this mean?

> All features that rely on using kvstore will not work until ...

We can't break kvstore mode for existing users. Is this intended just to mean that Users must not disable agents managing CIDs when using kvstore? Then maybe the text should say this.

Also for the implementation we should ensure that this incompatible configuration is reported to the user to ensure they don't end up breaking their environments.

Reply:
I completely agree with you. I think we shouldn't make operator CID management available to the users until we make it compatible with kvstore.

I updated the doc (at the top of the KVStore compatibility chapter) to state that this is a blocker for releasing the feature.